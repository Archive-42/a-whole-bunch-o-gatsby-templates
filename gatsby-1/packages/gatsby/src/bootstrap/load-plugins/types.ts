export interface IRawSiteConfig {
  plugins?: Array<PluginRef>
}

export interface ISiteConfig extends IRawSiteConfig {
  plugins?: Array<IPluginRefObject>
}

// There are two top-level "Plugin" concepts:
// 1. IPluginInfo, for processed plugins, and
// 2. PluginRef, for plugin configs

export interface IPluginInfo {
  /** Unique ID describing a plugin */
  id: string

  /** The absolute path to the plugin */
  resolve: string

  /** The plugin name */
  name: string

  /** The plugin version (can be content hash) */
  version: string

  /** Options passed to the plugin */
  pluginOptions?: IPluginInfoOptions
}

export interface IPluginInfoOptions {
  plugins?: Array<IPluginInfo>
  path?: string
  [option: string]: unknown
}

export interface IFlattenedPlugin extends IPluginInfo {
  skipSSR?: boolean
  ssrAPIs: Array<string>
  nodeAPIs: Array<string>
  browserAPIs: Array<string>
}

export interface IPluginRefObject {
  resolve: string
  options?: IPluginRefOptions
  parentDir?: string
}

export type PluginRef = string | IPluginRefObject

export interface IPluginRefOptions {
  plugins?: Array<PluginRef>
  path?: string
  [option: string]: unknown
}
