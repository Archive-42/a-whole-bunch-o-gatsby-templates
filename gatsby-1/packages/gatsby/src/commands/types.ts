import { PackageJson, Reporter } from "gatsby"
import { Store, AnyAction } from "redux"
import { IGatsbyState } from "../redux/types"

export interface ICert {
  key: Buffer
  cert: Buffer
}

export interface IDebugInfo {
  port: number
  break: boolean
}

export interface IProgram {
  _: `develop` | `build` | `clean` | `feedback` | `repl` | `serve`
  status?: string // I think this type should not exist here. It seems to be added in the reducer, but not applicable to the caller site from gatsby-cli
  useYarn: boolean
  open: boolean
  openTracingConfigFile: string
  port: number
  proxyPort: number
  host: string
  report: Reporter
  [`cert-file`]?: string
  [`key-file`]?: string
  directory: string
  https?: boolean
  sitePackageJson: PackageJson
  ssl?: ICert
  inspect?: number
  inspectBrk?: number
  graphqlTracing?: boolean
  verbose?: boolean
  setStore?: (store: Store<IGatsbyState, AnyAction>) => void
}

/**
 * @deprecated
 * Use `Stage` instead
 */

export enum BuildHTMLStage {
  DevelopHTML = `develop-html`,
  BuildHTML = `build-html`,
}

export enum Stage {
  Develop = `develop`,
  DevelopHTML = `develop-html`,
  BuildJavascript = `build-javascript`,
  BuildHTML = `build-html`,
}
