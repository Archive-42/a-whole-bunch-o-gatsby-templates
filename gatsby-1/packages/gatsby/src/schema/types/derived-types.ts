/**
 * Derived types are types that make sense only when their base type exists
 *
 * Take this node for example:
 * {
 *   internal: { type: 'Foo' },
 *   fields: {
 *     bar: "string",
 *   }
 * }
 *
 * It will produce following types:
 * Foo
 * FooFields
 *
 * FooInputFilter
 * FooSortInput
 *
 * FooFieldsInputFilter
 * FooFieldsSortFilter
 * etc
 *
 * Derived types:
 *   Foo: FooFields, FooInputFilter, FooSortInput
 *   FooFields: FooFieldsInputFilter, FooFieldsSortFilter
 *
 * Caveats:
 *   Only types created via inference are marked as derived. So if in the example above
 *   user explicitly defines `FooFields` type (via `createTypes` call) it won't be considered
 *   a derived type
 */
import {
  ObjectTypeComposer,
  InterfaceTypeComposer,
  ScalarTypeComposer,
  SchemaComposer,
  InputTypeComposer,
  EnumTypeComposer,
  UnionTypeComposer,
} from "graphql-compose"

type AllTypeComposer =
  | ObjectTypeComposer
  | InputTypeComposer
  | EnumTypeComposer
  | InterfaceTypeComposer
  | UnionTypeComposer
  | ScalarTypeComposer

const getDerivedTypes = ({
  typeComposer,
}: {
  typeComposer: AllTypeComposer
}): Set<string> => typeComposer.getExtension(`derivedTypes`) || new Set()

export const deleteFieldsOfDerivedTypes = ({ typeComposer }): void => {
  const derivedTypes = getDerivedTypes({ typeComposer })

  typeComposer.getFieldNames().forEach(fieldName => {
    const fieldType = typeComposer.getField(fieldName).type

    if (derivedTypes.has(fieldType.getTypeName())) {
      typeComposer.removeField(fieldName)
    }
  })
}

const removeTypeFromSchemaComposer = ({
  schemaComposer,
  typeComposer,
}): void => {
  schemaComposer.delete(typeComposer.getTypeName())
  schemaComposer.delete((typeComposer as any)._gqType)
  schemaComposer.delete(typeComposer)
}

export const clearDerivedTypes = ({
  schemaComposer,
  typeComposer,
}: {
  schemaComposer: SchemaComposer<any>
  typeComposer: AllTypeComposer
}): void => {
  const derivedTypes = getDerivedTypes({ typeComposer })

  for (const typeName of derivedTypes.values()) {
    const derivedTypeComposer = schemaComposer.getAnyTC(typeName)
    clearDerivedTypes({ schemaComposer, typeComposer: derivedTypeComposer })
    removeTypeFromSchemaComposer({
      schemaComposer,
      typeComposer: derivedTypeComposer,
    })
  }

  if (
    typeComposer instanceof ObjectTypeComposer ||
    typeComposer instanceof InterfaceTypeComposer
  ) {
    const inputTypeComposer = typeComposer.getInputTypeComposer()
    removeTypeFromSchemaComposer({
      schemaComposer,
      typeComposer: inputTypeComposer,
    })
    typeComposer.removeInputTypeComposer()
  }

  typeComposer.setExtension(`derivedTypes`, new Set())
}

export const addDerivedType = ({
  typeComposer,
  derivedTypeName,
}: {
  typeComposer: AllTypeComposer
  derivedTypeName: string
}): void => {
  const derivedTypes = getDerivedTypes({ typeComposer })
  typeComposer.setExtension(`derivedTypes`, derivedTypes.add(derivedTypeName))
}
