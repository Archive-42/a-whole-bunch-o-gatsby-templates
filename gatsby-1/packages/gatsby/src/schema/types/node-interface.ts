import { SORTABLE_ENUM } from "./sort"
import { SEARCHABLE_ENUM } from "./filter"
import {
  SchemaComposer,
  InterfaceTypeComposer,
  ObjectTypeComposer,
} from "graphql-compose"
import { IGatsbyResolverContext } from "../type-definitions"
import { Node } from "../../.."

export const NodeInterfaceFields = [`id`, `parent`, `children`, `internal`]

const getOrCreateNodeInterface = <TSource, TArgs>(
  schemaComposer: SchemaComposer<IGatsbyResolverContext<TSource, TArgs>>
): InterfaceTypeComposer => {
  // TODO: why is `mediaType` on Internal? Applies only to File!?
  // `fieldOwners` is an object
  // Should we drop ignoreType?
  const internalTC = schemaComposer.getOrCreateOTC(`Internal`, tc => {
    tc.addFields({
      content: `String`,
      contentDigest: `String!`,
      description: `String`,
      fieldOwners: [`String`],
      ignoreType: `Boolean`,
      mediaType: `String`,
      owner: `String!`,
      type: `String!`,
    })
    // TODO: Can be removed with graphql-compose 5.11
    tc.getInputTypeComposer()
  })

  const nodeInterfaceTC = schemaComposer.getOrCreateIFTC(`Node`, tc => {
    tc.setDescription(`Node Interface`)
    tc.addFields({
      id: `ID!`,
      parent: {
        type: `Node`,
        resolve: (source, _args, context): Node | null => {
          const { path } = context
          return context.nodeModel.getNodeById({ id: source.parent }, { path })
        },
        extensions: {
          searchable: SEARCHABLE_ENUM.SEARCHABLE,
          sortable: SORTABLE_ENUM.SORTABLE,
          needsResolve: true,
        },
      },
      children: {
        type: `[Node!]!`,
        resolve: (source, _args, context): Array<Node> => {
          const { path } = context
          return context.nodeModel.getNodesByIds(
            { ids: source.children },
            { path }
          )
        },
        extensions: {
          searchable: SEARCHABLE_ENUM.SEARCHABLE,
          sortable: SORTABLE_ENUM.SORTABLE,
          needsResolve: true,
        },
      },
      internal: internalTC.getTypeNonNull(),
    })
    // TODO: In Gatsby v2, the NodeInput.id field is of type String, not ID.
    // Remove this workaround for v3.
    const nodeInputTC = tc.getInputTypeComposer()
    nodeInputTC.extendField(`id`, { type: `String` })
  })

  return nodeInterfaceTC
}

export const addNodeInterfaceFields = <TSource = any, TArgs = any>({
  schemaComposer,
  typeComposer,
}: {
  schemaComposer: SchemaComposer<IGatsbyResolverContext<TSource, TArgs>>
  typeComposer: ObjectTypeComposer
}): void => {
  const nodeInterfaceTC = getOrCreateNodeInterface(schemaComposer)
  typeComposer.addFields(nodeInterfaceTC.getFields())
  nodeInterfaceTC.setResolveType(node => node.internal.type)
  schemaComposer.addSchemaMustHaveType(typeComposer)
}

export const addNodeInterface = <TSource = any, TArgs = any>({
  schemaComposer,
  typeComposer,
}: {
  schemaComposer: SchemaComposer<IGatsbyResolverContext<TSource, TArgs>>
  typeComposer: ObjectTypeComposer
}): void => {
  const nodeInterfaceTC = getOrCreateNodeInterface(schemaComposer)
  typeComposer.addInterface(nodeInterfaceTC)
  addNodeInterfaceFields({ schemaComposer, typeComposer })
}

export const getNodeInterface = <TSource = any, TContext = any, TArgs = any>({
  schemaComposer,
}: {
  schemaComposer: SchemaComposer<IGatsbyResolverContext<TSource, TArgs>>
}): InterfaceTypeComposer<TSource, TContext> =>
  getOrCreateNodeInterface(schemaComposer)
