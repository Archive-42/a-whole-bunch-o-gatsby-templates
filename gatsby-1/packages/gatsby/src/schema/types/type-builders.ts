import {
  ObjectTypeComposerAsObjectDefinition as ComposeObjectTypeConfig,
  InputTypeComposerAsObjectDefinition as ComposeInputObjectTypeConfig,
  InterfaceTypeComposerAsObjectDefinition as ComposeInterfaceTypeConfig,
  UnionTypeComposerAsObjectDefinition as ComposeUnionTypeConfig,
  EnumTypeComposerAsObjectDefinition as ComposeEnumTypeConfig,
  ScalarTypeComposerAsObjectDefinition as ComposeScalarTypeConfig,
} from "graphql-compose"

enum GatsbyGraphQLTypeKind {
  OBJECT = `OBJECT`,
  INPUT_OBJECT = `INPUT_OBJECT`,
  UNION = `UNION`,
  INTERFACE = `INTERFACE`,
  ENUM = `ENUM`,
  SCALAR = `SCALAR`,
}

export type GatsbyGraphQLType<TSource, TContext> =
  | {
      kind: GatsbyGraphQLTypeKind.OBJECT
      config: ComposeObjectTypeConfig<TSource, TContext>
    }
  | {
      kind: GatsbyGraphQLTypeKind.INPUT_OBJECT
      config: ComposeInputObjectTypeConfig
    }
  | {
      kind: GatsbyGraphQLTypeKind.UNION
      config: ComposeUnionTypeConfig<TSource, TContext>
    }
  | {
      kind: GatsbyGraphQLTypeKind.INTERFACE
      config: ComposeInterfaceTypeConfig<TSource, TContext>
    }
  | {
      kind: GatsbyGraphQLTypeKind.ENUM
      config: ComposeEnumTypeConfig
    }
  | {
      kind: GatsbyGraphQLTypeKind.SCALAR
      config: ComposeScalarTypeConfig
    }

function buildObjectType<TSource, TContext>(
  config: ComposeObjectTypeConfig<TSource, TContext>
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.OBJECT,
    config,
  }
}

function buildUnionType<TSource, TContext>(
  config: ComposeUnionTypeConfig<TSource, TContext>
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.UNION,
    config,
  }
}

function buildInterfaceType<TSource, TContext>(
  config: ComposeInterfaceTypeConfig<TSource, TContext>
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.INTERFACE,
    config,
  }
}

function buildInputObjectType<TSource, TContext>(
  config: ComposeInputObjectTypeConfig
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.INPUT_OBJECT,
    config,
  }
}

function buildEnumType<TSource, TContext>(
  config: ComposeEnumTypeConfig
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.ENUM,
    config,
  }
}

function buildScalarType<TSource, TContext>(
  config: ComposeScalarTypeConfig
): GatsbyGraphQLType<TSource, TContext> {
  return {
    kind: GatsbyGraphQLTypeKind.SCALAR,
    config,
  }
}

function isGatsbyType(something: any): something is GatsbyGraphQLTypeKind {
  return (
    typeof something === `object` &&
    something.kind &&
    GatsbyGraphQLTypeKind[something.kind]
  )
}

export {
  GatsbyGraphQLTypeKind,
  buildObjectType,
  buildUnionType,
  buildInterfaceType,
  buildInputObjectType,
  buildEnumType,
  buildScalarType,
  isGatsbyType,
}
