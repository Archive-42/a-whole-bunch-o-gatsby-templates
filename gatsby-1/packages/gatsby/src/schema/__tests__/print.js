const { build } = require(`..`)
import { buildObjectType } from "../types/type-builders"
const { store } = require(`../../redux`)
const { actions } = require(`../../redux/actions/restricted`)
const { actions: publicActions } = require(`../../redux/actions/public`)
const { createParentChildLink } = publicActions
const { printTypeDefinitions } = actions

jest.mock(`fs-extra`)
const fs = require(`fs-extra`)
afterEach(() => {
  fs.writeFile.mockClear()
})

jest.mock(`gatsby-cli/lib/reporter`, () => {
  return {
    log: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
    activityTimer: () => {
      return {
        start: jest.fn(),
        setStatus: jest.fn(),
        end: jest.fn(),
      }
    },
    phantomActivity: () => {
      return {
        start: jest.fn(),
        end: jest.fn(),
      }
    },
  }
})
const report = require(`gatsby-cli/lib/reporter`)
afterEach(() => {
  report.error.mockClear()
})

jest.spyOn(global.Date.prototype, `toISOString`).mockReturnValue(`2019-01-01`)

describe(`Print type definitions`, () => {
  beforeEach(() => {
    store.dispatch({ type: `DELETE_CACHE` })
    const node1 = {
      id: `test1`,
      internal: {
        type: `Test`,
      },
      children: [],
      foo: 26,
    }
    const node2 = {
      id: `test2`,
      parent: `test1`,
      internal: {
        type: `FooChild`,
      },
      bar: `bar`,
    }
    const node3 = {
      id: `test3`,
      parent: `test1`,
      internal: {
        type: `BarChild`,
      },
      bar: `bar`,
    }
    store.dispatch({ type: `CREATE_NODE`, payload: { ...node1 } })
    store.dispatch({ type: `CREATE_NODE`, payload: { ...node2 } })
    store.dispatch({ type: `CREATE_NODE`, payload: { ...node3 } })
    createParentChildLink({ parent: node1, child: node2 })
    createParentChildLink({ parent: node1, child: node3 })
    const typeDefs = []
    typeDefs.push(`
      type AnotherTest implements Node & ITest {
        nested: Nested
        date: Date @dateformat(formatString: "YYYY")
        hello(planet: String = "world", language: [Language!]!): String!
      }
      type Nested {
        baz: Boolean
      }
      input Language {
        lang: String
        hello: String
      }
      type OneMoreTest implements Node @infer {
        bar: Boolean
      }
      interface ITest @nodeInterface {
        id: ID!
        date: Date @dateformat(formatString: "YYYY")
      }
      interface ITest2 implements Node {
        id: ID!
        date: Date @dateformat(formatString: "YYYY")
      }
      union ThisOrThat = AnotherTest | OneMoreTest
    `)
    typeDefs.push(
      buildObjectType({
        name: `InlineTest`,
        fields: {
          first: `type Inline { foo: Nested }`,
          second: {
            type: `Date`,
            args: {
              bar: `input Baz { qux: Boolean }`,
            },
            extensions: {
              dateformat: {
                formatString: `MM/DD/YYYY`,
              },
            },
          },
        },
        interfaces: [`Node`, `ITest`],
        extensions: {
          infer: true,
          childOf: {
            types: [`OneMoreTest`],
          },
        },
      }),
      buildObjectType({
        name: `BarChild`,
        fields: {
          id: `ID!`,
        },
        interfaces: [`Node`],
        extensions: {
          childOf: {
            types: [`Test`],
          },
        },
      })
    )
    store.dispatch({
      type: `CREATE_TYPES`,
      payload: typeDefs[0],
      plugin: { name: `gatsby-plugin-test` },
    })
    store.dispatch({
      type: `CREATE_TYPES`,
      payload: typeDefs[1],
      plugin: { name: `gatsby-plugin-another-test` },
    })
    store.dispatch({
      type: `CREATE_TYPES`,
      payload: typeDefs[2],
      plugin: { name: `gatsby-plugin-another-test` },
    })
  })

  it(`saves type definitions to default file`, async () => {
    store.dispatch(printTypeDefinitions({}))
    await build({})
    expect(fs.writeFile.mock.calls[0][0]).toBe(`schema.gql`)
  })

  it(`saves type definitions to specified file`, async () => {
    store.dispatch(printTypeDefinitions({ path: `typedefs.graphql` }))
    await build({})
    expect(fs.writeFile.mock.calls[0][0]).toBe(`typedefs.graphql`)
  })

  it(`shows error message when file already exists`, async () => {
    fs.existsSync.mockImplementation(() => true)
    store.dispatch(printTypeDefinitions({ path: `typedefs.gql` }))
    await build({})
    expect(report.error).toHaveBeenCalledTimes(1)
    expect(report.error).toHaveBeenCalledWith(
      `Printing type definitions aborted. The file \`typedefs.gql\` already exists.`
    )
    fs.existsSync.mockReset()
  })

  it(`saves correct type definitions`, async () => {
    store.dispatch(printTypeDefinitions({}))
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })

  it(`allows specifying types to include (and includes interfaces and field types)`, async () => {
    store.dispatch(
      printTypeDefinitions({
        include: {
          types: [`AnotherTest`],
        },
      })
    )
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })

  it(`allows specifying types owned by plugins to include`, async () => {
    store.dispatch(
      printTypeDefinitions({
        include: {
          plugins: [`gatsby-plugin-another-test`],
        },
      })
    )
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })

  it(`allows specifying types to exclude`, async () => {
    store.dispatch(
      printTypeDefinitions({
        exclude: {
          types: [`InlineTest`],
        },
      })
    )
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })

  it(`allows specifying types owned by plugins to exclude`, async () => {
    store.dispatch(
      printTypeDefinitions({
        exclude: {
          plugins: [`gatsby-plugin-test`],
        },
      })
    )
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })

  it(`allows explicitly listing types to include without including field types`, async () => {
    store.dispatch(
      printTypeDefinitions({
        include: {
          types: [`InlineTest`],
        },
        withFieldTypes: false,
      })
    )
    await build({})
    expect(fs.writeFile.mock.calls[0][1]).toMatchSnapshot()
  })
})
