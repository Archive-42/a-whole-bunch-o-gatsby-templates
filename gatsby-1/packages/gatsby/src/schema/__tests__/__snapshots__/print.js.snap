// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Print type definitions allows explicitly listing types to include without including field types 1`] = `
"### Type definitions saved at 2019-01-01 ###

type InlineTest implements Node & ITest @childOf(types: [\\"OneMoreTest\\"]) @dontInfer {
  first: Inline
  second(bar: Baz): Date @dateformat(formatString: \\"MM/DD/YYYY\\")
}"
`;

exports[`Print type definitions allows specifying types owned by plugins to exclude 1`] = `
"### Type definitions saved at 2019-01-01 ###

type File implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Directory implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Site implements Node @dontInfer {
  buildTime: Date @dateformat
  siteMetadata: SiteSiteMetadata
}

type SiteSiteMetadata {
  title: String
  description: String
}

type SiteFunction implements Node @dontInfer {
  functionRoute: String!
  pluginName: String!
  originalAbsoluteFilePath: String!
  originalRelativeFilePath: String!
  relativeCompiledFilePath: String!
  absoluteCompiledFilePath: String!
  matchPath: String
}

type SitePage implements Node @dontInfer {
  path: String!
  component: String!
  internalComponentName: String!
  componentChunkName: String!
  matchPath: String
}

type Inline {
  foo: Nested
}

type Nested {
  baz: Boolean
}

input Baz {
  qux: Boolean
}

type InlineTest implements Node & ITest @childOf(types: [\\"OneMoreTest\\"]) @dontInfer {
  first: Inline
  second(bar: Baz): Date @dateformat(formatString: \\"MM/DD/YYYY\\")
}

interface ITest @nodeInterface {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

type BarChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}

type Test implements Node @dontInfer {
  foo: Int
}

type FooChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}"
`;

exports[`Print type definitions allows specifying types owned by plugins to include 1`] = `
"### Type definitions saved at 2019-01-01 ###

type InlineTest implements Node & ITest @childOf(types: [\\"OneMoreTest\\"]) @dontInfer {
  first: Inline
  second(bar: Baz): Date @dateformat(formatString: \\"MM/DD/YYYY\\")
}

interface ITest @nodeInterface {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

type Inline {
  foo: Nested
}

type Nested {
  baz: Boolean
}

input Baz {
  qux: Boolean
}

type BarChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}"
`;

exports[`Print type definitions allows specifying types to exclude 1`] = `
"### Type definitions saved at 2019-01-01 ###

type File implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Directory implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Site implements Node @dontInfer {
  buildTime: Date @dateformat
  siteMetadata: SiteSiteMetadata
}

type SiteSiteMetadata {
  title: String
  description: String
}

type SiteFunction implements Node @dontInfer {
  functionRoute: String!
  pluginName: String!
  originalAbsoluteFilePath: String!
  originalRelativeFilePath: String!
  relativeCompiledFilePath: String!
  absoluteCompiledFilePath: String!
  matchPath: String
}

type SitePage implements Node @dontInfer {
  path: String!
  component: String!
  internalComponentName: String!
  componentChunkName: String!
  matchPath: String
}

type AnotherTest implements Node & ITest @dontInfer {
  nested: Nested
  date: Date @dateformat(formatString: \\"YYYY\\")
  hello(planet: String = \\"world\\", language: [Language!]!): String!
}

interface ITest @nodeInterface {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

type Nested {
  baz: Boolean
}

input Language {
  lang: String
  hello: String
}

type OneMoreTest implements Node @dontInfer {
  bar: Boolean
}

interface ITest2 implements Node {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

union ThisOrThat = AnotherTest | OneMoreTest

type Inline {
  foo: Nested
}

input Baz {
  qux: Boolean
}

type BarChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}

type Test implements Node @dontInfer {
  foo: Int
}

type FooChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}"
`;

exports[`Print type definitions allows specifying types to include (and includes interfaces and field types) 1`] = `
"### Type definitions saved at 2019-01-01 ###

type AnotherTest implements Node & ITest @dontInfer {
  nested: Nested
  date: Date @dateformat(formatString: \\"YYYY\\")
  hello(planet: String = \\"world\\", language: [Language!]!): String!
}

interface ITest @nodeInterface {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

type Nested {
  baz: Boolean
}

input Language {
  lang: String
  hello: String
}"
`;

exports[`Print type definitions saves correct type definitions 1`] = `
"### Type definitions saved at 2019-01-01 ###

type File implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Directory implements Node @dontInfer {
  sourceInstanceName: String!
  absolutePath: String!
  relativePath: String!
  extension: String!
  size: Int!
  prettySize: String!
  modifiedTime: Date! @dateformat
  accessTime: Date! @dateformat
  changeTime: Date! @dateformat
  birthTime: Date! @dateformat
  root: String!
  dir: String!
  base: String!
  ext: String!
  name: String!
  relativeDirectory: String!
  dev: Int!
  mode: Int!
  nlink: Int!
  uid: Int!
  gid: Int!
  rdev: Int!
  ino: Float!
  atimeMs: Float!
  mtimeMs: Float!
  ctimeMs: Float!
  atime: Date! @dateformat
  mtime: Date! @dateformat
  ctime: Date! @dateformat
  birthtime: Date @deprecated(reason: \\"Use \`birthTime\` instead\\")
  birthtimeMs: Float @deprecated(reason: \\"Use \`birthTime\` instead\\")
}

type Site implements Node @dontInfer {
  buildTime: Date @dateformat
  siteMetadata: SiteSiteMetadata
}

type SiteSiteMetadata {
  title: String
  description: String
}

type SiteFunction implements Node @dontInfer {
  functionRoute: String!
  pluginName: String!
  originalAbsoluteFilePath: String!
  originalRelativeFilePath: String!
  relativeCompiledFilePath: String!
  absoluteCompiledFilePath: String!
  matchPath: String
}

type SitePage implements Node @dontInfer {
  path: String!
  component: String!
  internalComponentName: String!
  componentChunkName: String!
  matchPath: String
}

type AnotherTest implements Node & ITest @dontInfer {
  nested: Nested
  date: Date @dateformat(formatString: \\"YYYY\\")
  hello(planet: String = \\"world\\", language: [Language!]!): String!
}

interface ITest @nodeInterface {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

type Nested {
  baz: Boolean
}

input Language {
  lang: String
  hello: String
}

type OneMoreTest implements Node @dontInfer {
  bar: Boolean
}

interface ITest2 implements Node {
  id: ID!
  date: Date @dateformat(formatString: \\"YYYY\\")
}

union ThisOrThat = AnotherTest | OneMoreTest

type Inline {
  foo: Nested
}

input Baz {
  qux: Boolean
}

type InlineTest implements Node & ITest @childOf(types: [\\"OneMoreTest\\"]) @dontInfer {
  first: Inline
  second(bar: Baz): Date @dateformat(formatString: \\"MM/DD/YYYY\\")
}

type BarChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}

type Test implements Node @dontInfer {
  foo: Int
}

type FooChild implements Node @childOf(types: [\\"Test\\"]) @dontInfer {
  bar: String
}"
`;
